arelle/ValidateDuplicateFacts.py

```diff
@@ -0,0 +1,322 @@
+        if factA.concept.isLanguage and factB.concept.isLanguage and xValueA is not None and xValueB is not None:
+            return xValueA.lower() == xValueB.lower()  # required to handle case insensitivity
+        if isinstance(xValueA, DateTime):  # with/without time makes values unequal
+            return xValueA.dateOnly == cast(DateTime, xValueB).dateOnly and xValueA == xValueB
+        return xValueA == xValueB  # required to handle date/time with 24 hrs.
+    return factA.value == factB.value
+
+
+def getAspectEqualFacts(hashEquivalentFacts: list[ModelFact]) -> Iterator[list[ModelFact]]:
+    """
+    Given a list of concept/context/unit hash-equivalent facts,
+    yields sublists of aspect-equal facts from this list.
+    :param hashEquivalentFacts:
+    :return: Lists of aspect-equal facts.
+    """
+    aspectEqualFacts: dict[tuple[QName, str], dict[tuple[ModelContext, ModelUnit], list[ModelFact]]] = defaultdict(dict)
+    for fact in hashEquivalentFacts:  # check for hash collision by value checks on context and unit
+        contextUnitDict = aspectEqualFacts[(
+            fact.qname,
+            (fact.xmlLang or "").lower() if fact.concept.type.isWgnStringFactType else None
+        )]
+        _matched = False
+        for (context, unit), contextUnitFacts in contextUnitDict.items():
+            if fact.context is None:
+                if context is not None:
+                    continue
+            elif not fact.context.isEqualTo(context):
+                continue
+            if fact.unit is None:
+                if unit is not None:
+                    continue
+            elif not fact.unit.isEqualTo(unit):
+                continue
+            _matched = True
+            contextUnitFacts.append(fact)
+            break
+        if not _matched:
+            contextUnitDict[(fact.context, fact.unit)] = [fact]
+    for contextUnitDict in aspectEqualFacts.values():  # dups by qname, lang
+        for duplicateFacts in contextUnitDict.values():  # dups by equal-context equal-unit
+            if len(duplicateFacts) < 2:
+                continue
+            yield duplicateFacts
+
+
+def getDuplicateFactsOfType(factsInInstance: set[ModelFact], duplicateType: DuplicateType) -> Iterator[DuplicateFactPair]:
+    """
+    Yields each pairing of facts from the provided set that are duplicates of the given type(s).
+    :param factsInInstance:
+    :param duplicateType: Which types of duplicates should be returned.
+    :return: List of duplicate fact pairs.
+    """
+    if duplicateType == DuplicateType.NONE:
+        return
+    hashEquivalentFactGroups = getHashEquivalentFactGroups(factsInInstance)
+    for hashEquivalentFacts in hashEquivalentFactGroups:
+        if len(hashEquivalentFacts) < 2:
+            continue
+        for duplicateFactList in getAspectEqualFacts(hashEquivalentFacts):  # dups by equal-context equal-unit
+            for factA, factB in itertools.combinations(duplicateFactList, 2):
```

Comment:
I added an option, `firstDuplicateOnly`, to `getDuplicateFactsOfType` that I think addresses your concern in a fairly simple way. For validation, it will only return the first matching duplicate found for each group of duplicates. This is a better match to the current OIM implementation that would only fire one warning per group.

For *deduplication*, we'll have this flag set to false so we can address all duplicate pairs.

Reply:
I'm not sure it makes sense to consider "pairs", it seems like you always want to consider sets of facts.  If you have facts {A1,A2,B} where A1==A2 for whatever definition, users want to know the full set of facts that need to be consistent, not separate errors for {A1,B} and {A2,B}.  For "consistent" duplicates, the whole set of facts with matching concept/context/unit are either consistent or not (and you need to group by decimals, and values all need to match within a decimal group and be consistent across decimal groups).

Reply:
That clarifies the warning use case for me. I'll rework this PR to handle most evaluation as sets instead of pairs. For the deduplication feature that will be implemented next,  I expect that it will need more pair-level evaluation to capture all deduplication opportunities (that's what inspired this approach prematurely).

Reply:
> For the deduplication feature that will be implemented next, I expect that it will need more pair-level evaluation to capture all deduplication opportunities (that's what inspired this approach prematurely).

I still wouldn't expect pairs for that.

For non-numerics, I would expect something like `map<tuple<concept,context,lang(?)>,list<fact>>`, and select a single/first fact for each key.

For numerics, I would expect `map<tuple<concept,context,unit>,map<decimal,map<value,list<fact>>>>`.  If all decimal groups have a single value and all those values are consistent, then choose the most precise fact.  Otherwise, how much to deduplicate is a little unclear, but I'd suggest choosing a single/first fact for each value of each decimal group.